session-state-machine/src/main/scala/zio/raft/sessionstatemachine/SessionStateMachine.scala

```diff
@@ -0,0 +1,573 @@
+                sessionId,
+                newLastId
+              )
+
+              // Add all requests with composite keys
+              val stateWithRequests = requestsWithIds.foldLeft(stateWithNewId) {
+                case (s, (sid, requestId, pending)) =>
+                  s.updated["serverRequests"]((sid, requestId), pending)
+              }
+
+              // Create requests with context
+              val requestsWithContext = requestsWithIds.map { case (sid, requestId, pending) =>
+                ServerRequestWithContext(
+                  sessionId = sid,
+                  requestId = requestId,
+                  payload = pending.payload
+                )
+              }.toList
+
+              (stateWithRequests, accumulated ++ requestsWithContext)
+          }
+
+        (allRequestsWithContext, finalState)
+      }
+
+  /** Remove all session data when session expires.
+    */
+  private def expireSession(sessionId: SessionId): State[HMap[Schema], Unit] =
+    State.update { state =>
+      // Remove all cache entries for this session using range query
+      val cacheKeysToRemove = state.range["cache"](
+        (sessionId, RequestId.zero),
+        (sessionId, RequestId.max)
+      ).map(_._1)
+
+      // Remove all server requests for this session using range query
+      val serverRequestKeysToRemove = state.range["serverRequests"](
+        (sessionId, RequestId.zero),
+        (sessionId, RequestId.max)
+      ).map(_._1)
+
+      // Remove all session data in batch
+      state
+        .removedAll["cache"](cacheKeysToRemove)
+        .removedAll["serverRequests"](serverRequestKeysToRemove)
+        .removed["metadata"](sessionId)
+        .removed["lastServerRequestId"](sessionId)
+        .removed["highestLowestRequestIdSeen"](sessionId)
+    }
+
+  // ====================================================================================
+  // HELPER METHODS - All return State[HMap[Schema], A] for composition
+  // ====================================================================================
+
+  /** Clean up cached responses based on lowestRequestId (Lowest Sequence Number Protocol).
+    *
+    * Removes all cached responses for the session with requestId < lowestRequestId. This allows the client to control
+    * cache cleanup by telling the server which responses it no longer needs (Chapter 6.3 of Raft dissertation).
+    *
+    * Uses range queries to efficiently find and remove old cache entries.
+    */
+  private def cleanupCache(
+    sessionId: SessionId,
+    lowestRequestId: RequestId
+  ): State[HMap[Schema], Unit] =
+    State.update { state =>
+      // Use range to find all cache entries for this session with requestId < lowestRequestId
+      // Range is [from, until), so we use (sessionId, RequestId.zero) to (sessionId, lowestRequestId)
+      val keysToRemove = state.range["cache"](
+        (sessionId, RequestId.zero),
+        (sessionId, lowestRequestId)
```

Comment:
I think it should be lowestRequestId.next, no? as range is exclusive.

Please add a test as well, to make sure we properly clean up the cache